src/libPMacc/include/nvidia/atomic.hpp

```diff
@@ -0,0 +1,102 @@
+/**
+ * Copyright 2015 Rene Widera
+ *
+ * This file is part of libPMacc.
+ *
+ * libPMacc is free software: you can redistribute it and/or modify
+ * it under the terms of either the GNU General Public License or
+ * the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * libPMacc is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License and the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * and the GNU Lesser General Public License along with libPMacc.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+
+#include "types.h"
+#include "nvidia/warp.hpp"
+#include <math_functions.h>
+#include <device_functions.h>
+
+
+namespace PMacc
+{
+namespace nvidia
+{
+
+/** optimized atomic increment
+ *
+ * - only optimized if PTX ISA >=3.0
+ * - this atomic uses warp aggregation to speedup the operation compared to
+ *   cuda `atomicInc()`
+ * - cuda `atomicAdd()` is used if the compute architecture not supports
+ *   warp aggregation
+ *   is used
+ * - all participate threads must change the same
+ *   pointer (ptr) else the result is unspecified
+ *
+ * @param ptr pointer to memory (must be the same address for all threads in a block)
+ *
+ * This warp aggregated atomic increment implementation based on
+ * nvidia parallel forall example
+ * http://devblogs.nvidia.com/parallelforall/cuda-pro-tip-optimized-filtering-warp-aggregated-atomics/
+ */
```

Comment:
Pls add at least the author of the original post and date of the post.
Also you can write him on GitHub and ask for the license, too - or give him a call, he works in Jülich.
https://github.com/canonizer